<a href="https://colab.research.google.com/github/alejandra-leuno/Portfolio-AlejandraLeuno/blob/main/LEUNO_ALEJANDRA_TP_AA_2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Trabajo Practico Final

El objetivo de este trabajo práctico es analizar las particularidades de la utilización de
algoritmos para la generación de modelos de aprendizaje automático.

Fechas de entrega obligatoria:  
- 01/8/2023 a las 23:59 hs.

Importante: las entregas obligatorias, deben efectuarse a través del campus la UnaHur.

La entrega será individual, en formato de Nootebook. La cual debe presentar un formato prolijo con todas las notas explicativas que considere pertinentes. Al mismo tiempo, debe entregar un set de datos, producto de las transformaciones realizadas hasta el punto 4.

El nombre del archivo deberá respetar el siguiente formato: "APELLIDO_NOMBRE_TP_AA_2023.ipynb"

## Generar Data Frame

Previa conexion a Mi Drive:
- A partir del set de datos insurance_aa_2023.csv generar el data frame "data".

EL objetivo de este trabajo práctico será predecir el costo médico de personas en EE.UU. utilizando los distintos modelos vistos en el curso. Podrán encontrar más información sobre el dataset [aqui](https://www.kaggle.com/datasets/mirichoi0218/insurance?select=insurance.csv).

## Exploración del dataset

1. A partir de los datos entregados, describir los atributos realizando una breve explicación de qué representan y del tipo de variable (categórica, numérica u ordinal). En caso de que haya variables no numéricas, reportar los posibles valores que toman y cuán frecuentemente lo hacen.
  - Ver los primeros seis registros
  - Ver el tamaño del dataset
  - Ver los tipos de datos de cada columna
  - Usar el método *describe()* para ver las variables numéricas

2. Verificar la existencia de valores faltantes. Si los hay, ¿cuántos son y en qué atributos se encuentran? En caso de haberlos, realice la imputación de los mismos.

3. Estandarizar las variables numéricas (con excepción de _charges_ que es la variable a predecir).

4. Usar el metodo _get_dummies()_ para reemplazar las variables categóricas

## Modelos

5. Dividir el dataset en un conjunto de entrenamiento(90%) y uno de testeo (10%).

6. Realizar: (a) un modelo de regresión lineal; (b) modelo de árboles de decisión y (c) random forest. En todos los casos elija un hiperparámetro para ajustar utilizando **cross-validation**. ¿Cuáles fueron las variables más importantes para cada uno de los modelos?

7. Mencione la forma en la que se controló el overfitting en cada uno de los modelos realizados en el punto anterior.

8. Compare los modelos realizados. ¿Cuál considera que obtuvo un mejor resultado?

# Generar Data Frame

In [ ]:
# Montaje del Drive y carga de librerías generales
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np


Mounted at /content/drive


In [ ]:
# Cargar archivo CSV desde Google Drive
ruta_archivo = '/content/drive/MyDrive/Aprendizaje_Automatico/insurance_aa_2023.csv'
data = pd.read_csv(ruta_archivo)


# 1. Exploración del dataset



Variables definidas en el Dataset:
1. Age: variable numérica discreta que varía entre 18 y 64
2. Sex: variable categórica-dicotómica que se clasifica en Male y Female
3. BMI: variable numérica continua que oscila entre 15.96 y 53.13
4. Children: variable numérica discreta que varía entre 0 y 5
5. Smoker: variable categórica-dicotómica que se clasifica en Yes y No
6. Region: variable categórica, con 4 zonas posibles
7. Charges: variable numérica continua a predecir con los modelos


In [ ]:
# Ver los primeros seis registros
data.head (6)

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520
5   31  female  25.740         0     no  southeast   3756.62160

In [ ]:
# Ver tamaño del dataset
data.shape

(1338, 7)

In [ ]:
# Ver los tipos de datos de cada columna
data.info ()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1320 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


In [ ]:
# Usar el método describe() para ver las variables numéricas
data.describe ()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

# 2. Datos faltantes

In [ ]:
# Verificación de valores faltantes en el atributo Region (variable no numérica)
data.isna().sum()

age          0
sex          0
bmi          0
children     0
smoker       0
region      18
charges      0
dtype: int64

In [ ]:
# Se importa la clase SimpleImputer
from sklearn.impute import SimpleImputer

# Se crea un objeto SimpleImputer con la estrategia (most frequent) que es posible para esta variable no numérica (strings data)
# Para esta variable no numérica, no se puede aplicar media o mediana
imputer = SimpleImputer(missing_values = np.nan, strategy ='most_frequent')
col=data["region"].values.reshape(-1,1)

# Se ajustan los datos al objeto imputer = imputer.fit(col)
imputer = imputer.fit(col)

# Se imputan los datos faltantes
data['region']= pd.DataFrame(imputer.transform(col))

In [ ]:
# Verificación de datos completos en el atributo Region
data.isna().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

# 3. Estandarización variables numéricas

In [ ]:
def zscore(data):
  out = (data - data.mean())/data.std()
  return out

In [ ]:
# Se crea una lista con las columnas con datos tipo float64 e int64, con excepción de la variable 'charges'
col = ['age', 'bmi', 'children']

In [ ]:
data[col]

age     bmi  children
0      19  27.900         0
1      18  33.770         1
2      28  33.000         3
3      33  22.705         0
4      32  28.880         0
...   ...     ...       ...
1333   50  30.970         3
1334   18  31.920         0
1335   18  36.850         0
1336   21  25.800         0
1337   61  29.070         0

[1338 rows x 3 columns]

In [ ]:
# Estandarizar variables numéricas
from sklearn.preprocessing import StandardScaler

In [ ]:
# Se aplica la estandarización a las variables numéricas
scaler = StandardScaler()
#scaler = scaler.fit(data[col])
#data[col] = scaler.transform(data[col])
data[col] = scaler.fit_transform(data[col])

In [ ]:
# Comprobación de la estandarización de las variables numéricas
data.describe()

age           bmi      children       charges
count  1.338000e+03  1.338000e+03  1.338000e+03   1338.000000
mean  -1.805565e-16 -2.124194e-16 -5.576008e-17  13270.422265
std    1.000374e+00  1.000374e+00  1.000374e+00  12110.011237
min   -1.509965e+00 -2.412011e+00 -9.086137e-01   1121.873900
25%   -8.691547e-01 -7.164063e-01 -9.086137e-01   4740.287150
50%   -1.474046e-02 -4.320880e-02 -7.876719e-02   9382.033000
75%    8.396738e-01  6.611572e-01  7.510793e-01  16639.912515
max    1.765289e+00  3.685522e+00  3.240619e+00  63770.428010

In [ ]:
# Transformación variables categóricas
# Distribución y frecuencia de la Variable 'sex'
data['sex'].value_counts()

sex
male      676
female    662
Name: count, dtype: int64

In [ ]:
# Distribución y frecuencia de la Variable 'smoker'
data['smoker'].value_counts()

smoker
no     1064
yes     274
Name: count, dtype: int64

In [ ]:
# Reemplazo por valores numéricos
data['smoker'].map({'no':0, 'yes':1})

0       1
1       0
2       0
3       0
4       0
       ..
1333    0
1334    0
1335    0
1336    0
1337    1
Name: smoker, Length: 1338, dtype: int64

In [ ]:
# Reemplazo por valores numéricos
data['sex'].map({'female':0, 'male':1})

0       0
1       1
2       1
3       1
4       1
       ..
1333    1
1334    0
1335    0
1336    0
1337    0
Name: sex, Length: 1338, dtype: int64

# 4. Transformación de variables categóricas

In [ ]:
# Transformación de variables categóricas con el método Get_Dummies
data = pd.get_dummies(data, columns=['sex', 'smoker', 'region'], prefix=['sex', 'smoker','region'])

In [ ]:
# Revisión de columnas al final de la etapa de pre-procesamiento
data.head()

age       bmi  children      charges  sex_female  sex_male  smoker_no  \
0 -1.438764 -0.453320 -0.908614  16884.92400        True     False      False   
1 -1.509965  0.509621 -0.078767   1725.55230       False      True       True   
2 -0.797954  0.383307  1.580926   4449.46200       False      True       True   
3 -0.441948 -1.305531 -0.908614  21984.47061       False      True       True   
4 -0.513149 -0.292556 -0.908614   3866.85520       False      True       True   

   smoker_yes  region_northeast  region_northwest  region_southeast  \
0        True             False             False             False   
1       False             False             False              True   
2       False             False             False              True   
3       False             False              True             False   
4       False             False              True             False   

   region_southwest  
0              True  
1             False  
2             False  
3             False  
4             False

In [ ]:
# Revisión de los tipos de datos para las variables, al final de la etapa de pre-procesamiento
data.dtypes

age                 float64
bmi                 float64
children            float64
charges             float64
sex_female             bool
sex_male               bool
smoker_no              bool
smoker_yes             bool
region_northeast       bool
region_northwest       bool
region_southeast       bool
region_southwest       bool
dtype: object

In [ ]:
#Descargo el archivo from google.colab import files files.download('data.csv')
data.to_csv('data.csv')

# MODELOS

In [ ]:
# Importación de librerías
from sklearn.linear_model import RidgeCV
from sklearn.model_selection import train_test_split, KFold
from sklearn.model_selection import cross_val_score
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
import matplotlib.pyplot as plt


In [ ]:
# División del dataset en un conjunto de entrenamiento (90%) y un conjunto de testeo (10%)
X = data.drop('charges', axis = 1)
y = data['charges']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=42)

In [ ]:
print(data.shape)

(1338, 12)


In [ ]:
print("Tamaño del conjunto de entrenamiento:", X_train.shape)
print("Tamaño del conjunto de prueba:", X_test.shape)

Tamaño del conjunto de entrenamiento: (1204, 11)
Tamaño del conjunto de prueba: (134, 11)


# Regresión Lineal Ridge Cross Validation

In [ ]:
# Se divide el conjunto de entrenamiento en 4 folds.

# Se definen los valores de Alpha a probar
alphas = np.arange(0.01, 0.41, 0.01)

# Modelo Ridge
reg_ridgeCV =RidgeCV(alphas=alphas, cv=4)
reg_ridgeCV.fit(X_train, y_train)

RidgeCV(alphas=array([0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 , 0.11,
       0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 , 0.21, 0.22,
       0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 , 0.31, 0.32, 0.33,
       0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 ]),
        cv=4)

In [ ]:
# Predicciones en el conjunto de testeo para Ridge
y_test_pred_ridgeCV = reg_ridgeCV.predict(X_test)

# Cálculo del error cuadrático medio en el conjunto de testeo
mse_ridgeCV = mean_squared_error(y_test, y_test_pred_ridgeCV)

print("Mejor alpha encontrado:", reg_ridgeCV.alpha_)
print("Error cuadrático medio (RidgeCV):", mse_ridgeCV)

Mejor alpha encontrado: 0.4
Error cuadrático medio (RidgeCV): 32373465.86891563


In [ ]:
mse_alphas = []
alphas = np.arange(0.01, 0.41, 0.01)
for i in alphas:
  # Creación y entrenamiento del modelo de regresión con Ridge
  reg_ridgeCV = RidgeCV(alphas=i)
  reg_ridgeCV.fit(X_train, y_train)

  # Predicciones en el conjunto de predicción
  y_val_pred_ridgeCV = reg_ridgeCV.predict(X_test)

  # Cálculo del error cuadrático medio en el conjunto de testeo
  mse_ridgeCV = r2_score(y_test, y_val_pred_ridgeCV)
  mse_alphas.append(mse_ridgeCV)

  print("Error cuadrático medio (Ridge) con alpha = {} es {}".format(round(i, 2), round(mse_ridgeCV, 2)))

Error cuadrático medio (Ridge) con alpha = 0.01 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.02 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.03 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.04 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.05 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.06 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.07 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.08 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.09 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.1 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.11 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.12 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.13 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.14 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.15 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.16 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.17 es 0.77
Error cuadrático medio (Ridge) con alpha = 0.18 e

# Arbol de decisión

In [ ]:
# Módulos específicos para construcción y visualización del modelo
from sklearn.tree import DecisionTreeRegressor
from sklearn.tree import plot_tree

In [ ]:
# Construcción del árbol de decisión de profundidad 5
dtr = DecisionTreeRegressor(max_depth = 5, random_state = 42)

In [ ]:
# Entrenamiento del modelo con el método fit
dtr.fit(X_train, y_train)

DecisionTreeRegressor(max_depth=5, random_state=42)

In [ ]:
# Score del modelo en los datos de entrenamiento
score_in_train = dtr.score(X_train, y_train)

print('Score en el set de entrenamiento: {}'.format(score_in_train))

Score en el set de entrenamiento: 0.8826365386835131


In [ ]:
# Score del modelo en los datos de testeo
score_in_test = dtr.score(X_test, y_test)

print('Score en el set de testeo: {}'.format(score_in_test))

Score en el set de testeo: 0.8479272719559795


# Random Forests

In [ ]:
# Módulo especifíco para el modelo
from sklearn.ensemble import RandomForestRegressor

In [ ]:
# Construcción del modelo con 100 estimadores
# Por default, cada árbol individual es entrenado hasta máxima profundidad
rfr = RandomForestRegressor(n_estimators=100)

In [ ]:
# Entrenamiento del modelo
rfr.fit(X_train, y_train)

RandomForestRegressor()

In [ ]:
# Score en el set de entrenamiento
score_in_train = rfr.score(X_train, y_train)
print('Score en el set de train: {}'.format(score_in_train))

Score en el set de train: 0.9770259546008458


In [ ]:
# Score en el set de testeo
score_in_test = rfr.score(X_test, y_test)
print('Score en el set de testeo: {}'.format(score_in_test))

Score en el set de testeo: 0.845446576694149


In [ ]:
# Lista de features a considerar
features = ['age','bmi','children',	'sex_female','sex_male','smoker_no','smoker_yes','region_northeast','region_northwest','region_southeast','region_southwest']

In [ ]:
# Medición de la importancia de cada feature, donde las 3 variables que más se destacan son: Smoker, BMI y Age.
feature_importance = rfr.feature_importances_

print('Features: {}'.format(features))
print('Importancia = {}'.format(feature_importance))

Features: ['age', 'bmi', 'children', 'sex_female', 'sex_male', 'smoker_no', 'smoker_yes', 'region_northeast', 'region_northwest', 'region_southeast', 'region_southwest']
Importancia = [0.12973923 0.2072157  0.01903805 0.00310684 0.00348494 0.43388773
 0.1865465  0.00539175 0.00435028 0.00457261 0.00266636]


# CONCLUSIONES

1. **Regresión Lineal Ridge CV**

Bajo este modelo, dado un conjunto de alphas y una cantidad de 4 folds, el mejor alpha encontrado es de 0.4 y el score obtenido en el set de testeo es de 0.77.

2. **Árbol de decisión**

En este modelo, el score obtenido en el conjunto de entrenamiento es de 0.8826 y el score obtenido en el conjunto de testeo es de 0.8479. La diferencia de 0.0347 indica que los valores son muy similares y eso podría suponer que el overfitting fue controlado.

3. **Random Forests**

En este modelo, el score obtenido en el conjunto de entrenamiento es de 0.9765 y el score obtenido en el conjunto de testeo es de 0.8455. La diferencia de 0.131 podría suponer que el overfitting no fue controlado en el set de train.

De los tres modelos presentados, el mejor score obtenido en el conjunto de testeo corresponde al modelo de Árbol de Decisión.